pkg/runtime/unstructured/converter.go

```diff
@@ -232,3 +232,176 @@ func (c *Converter) structFromUnstructured(sv, dv reflect.Value) error {
 func (c *Converter) interfaceFromUnstructured(sv, dv reflect.Value) error {
 	return fmt.Errorf("Interface conversion unsupported")
 }
+
+var (
+	marshalerType          = reflect.TypeOf(new(encodingjson.Marshaler)).Elem()
+	mapStringInterfaceType = reflect.TypeOf(map[string]interface{}{})
+)
+
+func (c *Converter) ToUnstructured(obj runtime.Object, u *map[string]interface{}) error {
+	return c.toUnstructured(reflect.ValueOf(obj).Elem(), reflect.ValueOf(u).Elem())
+}
+
+func (c *Converter) toUnstructured(sv, dv reflect.Value) error {
+	st, dt := sv.Type(), dv.Type()
+
+	// Check if the object has a custom JSON marshaller/unmarshaller.
+	if st.Implements(marshalerType) {
+		var data []byte
+		var err error
+		if sv.Kind() == reflect.Ptr && sv.IsNil() {
+			data = []byte("null")
+		} else {
+			marshaler := sv.Interface().(encodingjson.Marshaler)
+			data, err = marshaler.MarshalJSON()
+			if err != nil {
+				return err
+			}
+		}
+		if string(data) == "null" {
+			// We're done - we don't need to store anything.
+		} else {
+			if len(data) > 0 && data[0] == '"' {
+				var result string
+				err := json.Unmarshal(data, &result)
+				if err != nil {
+					return fmt.Errorf("Error decoding from json: %v", err)
+				}
+				dv.Set(reflect.ValueOf(result))
+			} else if len(data) > 0 && data[0] == '{' {
+				result := make(map[string]interface{})
+				err := json.Unmarshal(data, &result)
+				if err != nil {
+					return fmt.Errorf("Error decoding from json: %v", err)
+				}
+				dv.Set(reflect.ValueOf(result))
+			} else {
+				var result int64
+				err := json.Unmarshal(data, &result)
+				if err != nil {
+					return fmt.Errorf("Error decoding from json: %v", err)
+				}
+				dv.Set(reflect.ValueOf(result))
+			}
+		}
+		return nil
+	}
+
+	switch st.Kind() {
+	case reflect.String, reflect.Bool,
+		reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64,
+		reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		if dt.Kind() == reflect.Interface && dv.NumMethod() == 0 {
+			dv.Set(reflect.New(st))
+		}
+		dv.Set(sv)
+		return nil
+	case reflect.Map:
+		return c.mapToUnstructured(sv, dv)
+	case reflect.Slice:
+		return c.sliceToUnstructured(sv, dv)
+	case reflect.Ptr:
+		return c.pointerToUnstructured(sv, dv)
+	case reflect.Struct:
+		return c.structToUnstructured(sv, dv)
+	case reflect.Interface:
+		return c.interfaceToUnstructured(sv, dv)
+	default:
+		return fmt.Errorf("unrecognized type: %v", st.Kind())
+	}
+}
+
+func (c *Converter) mapToUnstructured(sv, dv reflect.Value) error {
+	st, dt := sv.Type(), dv.Type()
+	if dt.Kind() == reflect.Interface && dv.NumMethod() == 0 {
+		dv.Set(reflect.MakeMap(mapStringInterfaceType))
+		dv = dv.Elem()
+		dt = dv.Type()
+	}
+	if dt.Kind() != reflect.Map {
+		return fmt.Errorf("Cannot convert struct to: %v", dt.Kind())
+	}
+
+	if !st.Key().AssignableTo(dt.Key()) && !st.Key().ConvertibleTo(dt.Key()) {
```

Comment:
Probably should cache and check this elsewhere - do a feasibility check for each st -> dt and cache the results.  This gets expensive.